shuttle/src/future/batch_semaphore.rs

```diff
@@ -752,3 +790,50 @@ impl crate::annotations::WithName for BatchSemaphore {
         self
     }
 }
+
+impl BatchSemaphore {
+    #[cfg(test)]
+    pub(crate) fn signature(&self) -> &ResourceSignature {
+        &self.signature
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn unique_resource_signature_batch_semaphore() {
+        crate::check_random(
+            || {
+                let sem1 = BatchSemaphore::new(1, Fairness::Unfair);
+                let sem2 = BatchSemaphore::new(1, Fairness::Unfair);
+                assert_ne!(sem1.signature, sem2.signature);
+            },
+            1,
+        );
+    }
+
+    #[test]
+    fn batch_semaphore_signatures_consistent_across_shuttle_iterations() {
+        use std::collections::HashSet;
+        use std::sync::{Arc, Mutex};
+
+        let all_signatures = Arc::new(Mutex::new(HashSet::new()));
+        let all_signatures_clone = all_signatures.clone();
+
+        crate::check_random(
+            move || {
+                let sem1 = BatchSemaphore::new(1, Fairness::StrictlyFair);
+                let sem2 = BatchSemaphore::new(2, Fairness::Unfair);
+
+                all_signatures_clone.lock().unwrap().insert(sem1.signature().clone());
+                all_signatures_clone.lock().unwrap().insert(sem2.signature().clone());
+            },
+            10,
+        );
+
+        // Should have exactly 2 unique signatures across all iterations
```

Comment:
Could we assert the stronger condition that sem1 and sem2 keep their signatures across iterations?  Could we store (number_of_available_permits, signature) in the HashSet instead?  Since you initialize them with different #permits?